examples/java/src/test/java/org/apache/beam/examples/WordCountIT.java

```diff
@@ -46,9 +67,92 @@
   public void testE2EWordCount() throws Exception {
     PipelineOptionsFactory.register(WordCountITOptions.class);
     WordCountITOptions options = TestPipeline.testingPipelineOptions().as(WordCountITOptions.class);
-    options.setOutput(Joiner.on("/").join(new String[]{options.getTempRoot(),
-        String.format("WordCountIT-%tF-%<tH-%<tM-%<tS-%<tL", new Date()), "output", "results"}));
+
+    options.setOutput(IOChannelUtils.resolve(
+            options.getTempRoot(),
+            String.format("WordCountIT-%tF-%<tH-%<tM-%<tS-%<tL", new Date()),
+            "output",
+            "results"));
+    options.setOnSuccessMatcher(new WordCountOnSuccessMatcher(options.getOutput() + "*"));
 
     WordCount.main(TestPipeline.convertToArgs(options));
   }
+
+  /**
+   * Matcher for verifying WordCount output data.
+   */
+  static class WordCountOnSuccessMatcher extends TypeSafeMatcher<PipelineResult>
+      implements SerializableMatcher<PipelineResult> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(WordCountOnSuccessMatcher.class);
+
+    private static final String EXPECTED_CHECKSUM = "0723901e2878b2913c548b3a497b9c86b71a5708";
+    private String actualChecksum;
+
+    private final String outputPath;
+
+    WordCountOnSuccessMatcher(String outputPath) {
+      this.outputPath = checkNotNull(outputPath);
+    }
+
+    @Override
+    protected boolean matchesSafely(PipelineResult pResult) {
+      if (outputPath.isEmpty()) {
+        fail(String.format("Expected valid output path, but received %s", outputPath));
+      }
+
+      try {
+        // Load output data
+        List<String> outputs = readLines(outputPath);
+
+        // Verify outputs. Checksum is computed using SHA-1 algorithm
+        Collections.sort(outputs);
+        actualChecksum =
+                Hashing.sha1().hashString(outputs.toString(), StandardCharsets.UTF_8).toString();
+        LOG.info("Generate checksum for output data: {}", actualChecksum);
+
+        return actualChecksum.equals(EXPECTED_CHECKSUM);
+      } catch (IOException e) {
+        throw new RuntimeException(
+                String.format("Fail to read from path: %s", outputPath));
+      }
+    }
+
+    private List<String> readLines(String path) throws IOException {
+      List<String> readData = new ArrayList<>();
+
+      IOChannelFactory factory = IOChannelUtils.getFactory(path);
+
+      // Match inputPath which may contains glob
+      Collection<String> files = factory.match(path);
+
+      // Read data from file paths
+      for (String file : files) {
+        try (Reader reader =
+                     Channels.newReader(factory.open(file), StandardCharsets.UTF_8.name())) {
+          readData.addAll(CharStreams.readLines(reader));
+        }
+      }
+      LOG.info("Read {} lines from {} files with destination path {}",
```

Comment:
List the actual files that were read from so that a person knows which ones to take a look at.